keras_nlp/src/models/deeplab_v3_plus/deeplab_v3_plus_segmenter.py

```diff
@@ -0,0 +1,237 @@
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+import keras
+
+from keras_nlp.src.api_export import keras_nlp_export
+from keras_nlp.src.models.deeplab_v3_plus.deeplab_v3_plus_layers import (
+    SpatialPyramidPooling,
+)
+from keras_nlp.src.models.segmentation import Segmentation
+
+
+@keras_nlp_export(
+    [
+        "keras_nlp.models.DeepLabV3Plus",
+        "keras_nlp.models.segmentation.DeepLabV3Plus",
+    ]
+)
+class DeepLabV3Plus(Segmentation):
+    """DeepLabV3+ architecture for semantic segmentation.
+
+    This class implements a DeepLabV3+ architecture as described in
+    [Encoder-Decoder with Atrous Separable Convolution for Semantic Image
+    Segmentation](https://arxiv.org/abs/1802.02611)(ECCV 2018)
+    and [Rethinking Atrous Convolution for Semantic Image Segmentation](
+        https://arxiv.org/abs/1706.05587)(CVPR 2017)
+
+    Args:
+    backbone: `keras.Model`. The backbone network for the model that is
+        used as a feature extractor for the DeepLabV3+ Encoder. Should
+        either be a `keras_nlp.models.backbones.backbone.Backbone` or a
+        `keras.Model` that implements the `pyramid_outputs`
+        property with keys "P2", "P3" etc as values. A
+        somewhat sensible backbone to use in many cases is the
+        `keras_nlp.models.ResNetBackbone.from_preset("resnet_v2_50")`.
+    num_classes: int, the number of classes for the detection model. Note
+        that the `num_classes` contains the background class, and the
+        classes from the data should be represented by integers with range
+        [0, `num_classes`).
+    projection_filters: int, number of filters in the convolution layer
+        projecting low-level features from the `backbone`.
+    low_level_feature_key: str, layer level to extract the feature from one of the
+    key from the `backbone` `pyramid_outputs`
+        property such as  "P2", "P3" etc.
+    spatial_pyramid_pooling_key: str, layer level to extract and perform
+    `spatial_pyramid_pooling`, one of the key from the `backbone` `pyramid_outputs`
+        property such as  "P4", "P5" etc.
+    spatial_pyramid_pooling: (Optional) a `keras.layers.Layer`. Also known
+        as Atrous Spatial Pyramid Pooling (ASPP). Performs spatial pooling
+        on different spatial levels in the pyramid, with dilation. If
+        provided, the feature map from the backbone is passed to it inside
+        the DeepLabV3 Encoder, otherwise SpatialPyramidPooling layer is used.
+    dilation_rates: (Optional) A `list` of integers for parallel dilated conv.
+        Applied only when Default `SpatialPyramidPooling` is used. Usually a
+        sample choice of rates are [6, 12, 18].
+    segmentation_head: (Optional) a `keras.layers.Layer`. If provided, the
```

Comment:
I think we actually don't want to do this. We are someone messing with the definitely of a CV backbone, but in general we have a pattern where...

* A backbone contains *all* the weights from a pretrained model.
* A task wraps the backbone for a specific task.

So with that approach, I think what we probably need to do here...

* Create a new `DeeplabV3PlusBackbone` class which has almost all the arguments here. Rename the `backbone` argument it takes in to a `feature_pyramid_model` or something like that.
* The `ImageSegmenter` wraps this backbone with a head (that is not configurable!). The arguments that this class takes should be the same as our classification tasks, `backbone`, `num_classes`, and `activation`.
* Anyone wanting a custom `segmentation_head` in the current design would instead create a `DeeplabV3PlusBackbone` and attach their own head to it.

So basically, the new `DeeplabV3PlusBackbone` will have weights from the pooling layers. On disk we will save something like this...

```python
config.json # DeeplabV3PlusBackbone
model.weights # DeeplabV3PlusBackbone weights (feature pyramid + pooling weights)
task.json # DeeplabV3PlusImageSegmenter config
task.weights # Head weights only
```

We will need to add a way to allow loading either the segmentation head with it's 19 classes, or a new head with random weights, but that does not need to be on this PR.

cc @divyashreepathihalli

Reply:
Made changes according to this